File: androidApp/src/main/kotlin/zinqq/app/theme/ZinqqTypography.kt
package zinqq.app.theme

import androidx.compose.material3.Typography
import androidx.compose.runtime.Immutable
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import zinqq.app.R

/**
 * Bundled OFL fonts matching the PWA's self-hosted set (U13, KTD-11, R12):
 * Inter 400–700 as `--font-sans`, Space Grotesk 500–700 as `--font-display`.
 * The TTFs are the exact files the PWA serves from `public/fonts/`, copied
 * into `res/font/` (no Google Fonts CDN, same privacy stance).
 */
val InterFamily = FontFamily(
    Font(R.font.inter_400, FontWeight.Normal),
    Font(R.font.inter_500, FontWeight.Medium),
    Font(R.font.inter_600, FontWeight.SemiBold),
    Font(R.font.inter_700, FontWeight.Bold),
)

val SpaceGroteskFamily = FontFamily(
    Font(R.font.space_grotesk_500, FontWeight.Medium),
    Font(R.font.space_grotesk_600, FontWeight.SemiBold),
    Font(R.font.space_grotesk_700, FontWeight.Bold),
)

/** The two font roles the PWA exposes as CSS variables. */
@Immutable
data class ZinqqFonts(
    /** `--font-sans`: body copy, labels, values. */
    val sans: FontFamily = InterFamily,
    /** `--font-display`: amounts, pills, headings, numpad. */
    val display: FontFamily = SpaceGroteskFamily,
)

/** Material typography with Inter as the app-wide default family. */
internal fun zinqqMaterialTypography(): Typography {
    val default = Typography()
    fun withSans(style: androidx.compose.ui.text.TextStyle) =
        style.copy(fontFamily = InterFamily)
    return Typography(
        displayLarge = withSans(default.displayLarge),
        displayMedium = withSans(default.displayMedium),
        displaySmall = withSans(default.displaySmall),
        headlineLarge = withSans(default.headlineLarge),
        headlineMedium = withSans(default.headlineMedium),
        headlineSmall = withSans(default.headlineSmall),
        titleLarge = withSans(default.titleLarge),
        titleMedium = withSans(default.titleMedium),
        titleSmall = withSans(default.titleSmall),
        bodyLarge = withSans(default.bodyLarge),
        bodyMedium = withSans(default.bodyMedium),
        bodySmall = withSans(default.bodySmall),
        labelLarge = withSans(default.labelLarge),
        labelMedium = withSans(default.labelMedium),
        labelSmall = withSans(default.labelSmall),
    )
}

File: iosApp/iosApp/Theme/ZinqqTypography.swift
import SwiftUI

/// Bundled OFL fonts matching the PWA's self-hosted set (U18, KTD-11, R12):
/// Inter 400–700 as `--font-sans`, Space Grotesk 500–700 as `--font-display`.
/// The TTFs are the exact files Android bundles in `res/font/`, copied into
/// `iosApp/Fonts/` (no system fallback for the brand roles, same privacy
/// stance as the PWA's self-hosting).
///
/// Registration mechanism: the TTFs ride into the bundle as plain resources
/// (XcodeGen picks them up from the `iosApp` sources dir) and `UIAppFonts`
/// lists them in the Info.plist written by project.yml's `info:` block —
/// see project.yml for how that merges with GENERATE_INFOPLIST_FILE.
enum ZinqqFont {
    /// `--font-sans`: body copy, labels, values. PostScript names from the
    /// bundled TTFs (Inter-Regular/Medium/SemiBold/Bold).
    static func sans(_ size: CGFloat, weight: Font.Weight = .regular) -> Font {
        .custom(interName(for: weight), size: size)
    }

    /// `--font-display`: amounts, pills, headings, numpad. Space Grotesk ships
    /// 500–700 only; lighter weights clamp to Medium.
    static func display(_ size: CGFloat, weight: Font.Weight = .medium) -> Font {
        .custom(spaceGroteskName(for: weight), size: size)
    }

    private static func interName(for weight: Font.Weight) -> String {
        switch weight {
        case .bold, .heavy, .black: return "Inter-Bold"
        case .semibold: return "Inter-SemiBold"
        case .medium: return "Inter-Medium"
        default: return "Inter-Regular"
        }
    }

    private static func spaceGroteskName(for weight: Font.Weight) -> String {
        switch weight {
        case .bold, .heavy, .black: return "SpaceGrotesk-Bold"
        case .semibold: return "SpaceGrotesk-SemiBold"
        default: return "SpaceGrotesk-Medium"
        }
    }
}
